src/Storages/ObjectStorageQueue/ObjectStorageQueueUnorderedFileMetadata.cpp

```diff
@@ -103,29 +103,45 @@ void ObjectStorageQueueUnorderedFileMetadata::setProcessedImpl()
     /// In one zookeeper transaction do the following:
     enum RequestType
     {
-        SET_MAX_PROCESSED_PATH = 0,
-        CHECK_PROCESSING_ID_PATH = 1, /// Optional.
-        REMOVE_PROCESSING_ID_PATH = 2, /// Optional.
-        REMOVE_PROCESSING_PATH = 3, /// Optional.
+        CHECK_PROCESSING_ID_PATH,
+        REMOVE_PROCESSING_ID_PATH,
+        REMOVE_PROCESSING_PATH,
+        SET_PROCESSED_PATH,
     };
 
     const auto zk_client = getZooKeeper();
-    std::string failure_reason;
-
     Coordination::Requests requests;
-    requests.push_back(
-        zkutil::makeCreateRequest(
-            processed_node_path, node_metadata.toString(), zkutil::CreateMode::Persistent));
+    std::map<RequestType, UInt8> request_id;
 
     if (processing_id_version.has_value())
     {
         requests.push_back(zkutil::makeCheckRequest(processing_node_id_path, processing_id_version.value()));
         requests.push_back(zkutil::makeRemoveRequest(processing_node_id_path, processing_id_version.value()));
         requests.push_back(zkutil::makeRemoveRequest(processing_node_path, -1));
+
+        /// The order is important:
+        /// we must first check processing nodes and set processed_path the last.
```

Comment:
Explanation for this fix is in this comment in code, but in more detail here:
The issue for this fix was about Unordered mode, but the same problem was with Ordered mode.
The point is that the order of keeper requests was mistaken, because if the first request failed - all later are automatically failed, but here it is ok that certain requests fail - it is expected behaviour if keeper session expired in the middle of processing, so the processing will be retried. So by mistake the non-retriable bug-indicating request was the first in the list of requests, while it must be the last, as it is expected that it fails if those requests which indicate retriable problem fail, which is expected and is not a logical error.